src/runtime/virtcontainers/utils/utils.go

```diff
@@ -390,3 +391,61 @@ outer:
 
 	return nil
 }
+
+func MkdirAllWithInheritedOwner(path string, perm os.FileMode) error {
+	var sb strings.Builder
+	sb.WriteString(path)
+	// Append '/' to the end if the path doesn't end with '/'
+	if !strings.HasSuffix(path, string(os.PathSeparator)) {
+		sb.WriteRune(os.PathSeparator)
+	}
+	newPath := sb.String()
+	var stack []int
+	for i := len(newPath) - 1; i > 0; i-- {
+		if os.IsPathSeparator(newPath[i]) {
+			stack = append(stack, i)
+		}
+	}
+
+	// By default, use the uid and gid of the calling process.
+	var uid = os.Getuid()
+	var gid = os.Getgid()
+
+	for j := len(stack) - 1; j >= 0; j-- {
+		curPath := newPath[:stack[j]]
+		info, err := os.Stat(curPath)
+		if err == nil {
+			if !info.IsDir() {
+				return &os.PathError{Op: "mkdir", Path: newPath, Err: syscall.ENOTDIR}
+			}
+			if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+				uid = int(stat.Uid)
+				gid = int(stat.Gid)
+			} else {
+				return fmt.Errorf("fail to retrieve the uid and gid of path %s", curPath)
+			}
+		} else {
+			if err = os.Mkdir(curPath, perm); err != nil {
+				return fmt.Errorf("mkdir call failed: %v", err.Error())
+			}
+			if err = syscall.Chown(curPath, uid, gid); err != nil {
+				return fmt.Errorf("chown syscall failed: %v", err.Error())
+			}
+		}
+	}
+	return nil
+}
+
+func ChownToParent(path string) error {
+	info, err := os.Stat(filepath.Dir(path))
+	if err != nil {
+		return fmt.Errorf("parent directory does not exist %s", filepath.Dir(path))
```

Comment:
I'd consider adding the original error to the returned error since this message _may_ be misleading in certain scenarios.